pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java

```diff
@@ -51,74 +51,84 @@ public RangeIndexBasedFilterOperator(PredicateEvaluator rangePredicateEvaluator,
 
   @Override
   protected FilterBlock getNextBlock() {
-    RangeIndexReader rangeIndexReader = (RangeIndexReader) _dataSource.getRangeIndex();
+    @SuppressWarnings("unchecked")
+    RangeIndexReader<ImmutableRoaringBitmap> rangeIndexReader =
+        (RangeIndexReader<ImmutableRoaringBitmap>) _dataSource.getRangeIndex();
     assert rangeIndexReader != null;
 
+    ImmutableRoaringBitmap matches;
+    // if the implementation cannot match the entire query exactly, it will
+    // yield partial matches, which need to be verified by scanning. If it
+    // can answer the query exactly, this will be null.
+    ImmutableRoaringBitmap partialMatches;
     int firstRangeId;
     int lastRangeId;
     if (_rangePredicateEvaluator instanceof SortedDictionaryBasedRangePredicateEvaluator) {
-      firstRangeId = rangeIndexReader
-          .findRangeId(((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getStartDictId());
       // NOTE: End dictionary id is exclusive in OfflineDictionaryBasedRangePredicateEvaluator.
-      lastRangeId = rangeIndexReader
-          .findRangeId(((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getEndDictId() - 1);
+      matches = rangeIndexReader.getMatchingDocIds(
+          ((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getStartDictId(),
```

Comment:
(nit) Cache the `min` and `max` value

Reply:
Why?

Reply:
Both performance and readability. What I mean is to put `((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getStartDictId()` and `((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getEndDictId() - 1` into local variables instead of calculating them twice

Reply:
There is no performance benefit to the suggested change, these values are constants in `SortedDictionaryBasedRangePredicateEvaluator` (and all other implementations used in this block of code):

```java
    public int getStartDictId() {
      return _startDictId;
    }

    public int getEndDictId() {
      return _endDictId;
    }
```

I will make the change on the basis of readability, but it's very difficult to spot bottlenecks by code review so it's better to rely on data.